### ISSUE TITLE
PKG Install related bugs....

### ISSUE THREAD

**Issue by remlei** [LightningMods/Itemzflow#9]
1st bug:

Fail to install any PKG files for the 2nd time that was interrupted during install process. Reinstalling Itemzflow doesnt fix the issue

how to reproduce: Just exit and close itemzflow during pkg install.

logs: 

2023-02-10 13:03:46 INFO  source/GLES2_filemanager.cpp:81: [FS_DEBUG] b4 rela_path: /hostapp
2023-02-10 13:03:46 INFO  source/GLES2_filemanager.cpp:100: [FS_DEBUG] after rela_path: /hostapp/test.pkg
2023-02-10 13:03:46 INFO  source/GLES2_filemanager.cpp:81: [FS_DEBUG] b4 rela_path: /hostapp
2023-02-10 13:03:46 INFO  source/GLES2_filemanager.cpp:100: [FS_DEBUG] after rela_path: /hostapp/test.pkg
2023-02-10 13:03:46 INFO  source/GLES2_itemzflow.cpp:1074: SETTING_V_CURR 9
2023-02-10 13:03:46 INFO  source/installpkg.cpp:288: Initializing BGFT...
2023-02-10 13:03:46 INFO  source/installpkg.cpp:315: pkginstall 1
2023-02-10 13:03:46 WARN  source/utils.cpp:1832: Install Failed with code
HEX: -0x7f66ff7a Int: -2137456506
sceBgftServiceIntDownloadRegisterTaskByStorageEx

2nd bug"

Bug when installing the a update file when the base game is not installed ended up having the pkg file lost somewhere in the ps4 and unable to install the update again regardless if im installing it via debug setting or or itemzlflow.

how to trigger: just install the update without the base game installed

-----

both bug will also result to losing the disk space used and no obvious way to recover the lost disc space other than initializing the hdd.

**Comment by LightningMods** [comment]
> Fail to install any PKG files for the 2nd time that was interrupted during install process. Reinstalling Itemzflow doesnt fix the issue

I had that happen before but I only had to cancel the install via the notification center for it to return to normal that error you get and is in the log is `SCE_BGFT_ERROR_CONTENT_ALREADY_DOWNLOADING` which happens because Itemzflow is installing the PKG so when you close the app it stops installing and cant continue so you have to cancel the install, its not really a bug but a PS4 limitation, that it doesn't continue the install if you close the app.

> Bug when installing the a update file when the base game is not installed ended up having the pkg file lost somewhere in the ps4 and unable to install the update again regardless if im installing it via debug setting or or itemzlflow.

I can add a check to see if the PKG is a patch and if it is then see if the game is installed

**Comment by LightningMods** [comment]
Hello, as we discussed before the first issue can't be fixed however I have implemented a check for the 2nd issue and tested it as working, I'll keep this open until I release the next small update soon-ish

here's a picture of the check in action (tested by trying to install patch without the game installed)

![image](https://user-images.githubusercontent.com/34066913/218281914-fd8c70ac-4441-4e6a-b787-897d7fc9c402.png)

**Comment by remlei** [comment]
Thats weird, should I submit you a link (to a fpkg game update of someone else) that installs the update and do nothing after reaching 100%, not even a notification popup. Normally installing a fpkg update without base game should be fine, it just sits in notification area stating ready to install, which I can do later after installing the base pkg. But in this case, thats not what happend.

**Comment by chris2677** [comment]
I have confirmed this issue, i have tried 3 different games and they all start but eventually freeze at different points.  I confirmed my firewall and antivirus is not the problem.  It seems the PS4 is freezing and when this happens you can not close the app or the PS4 will crash and reboot.

**Comment by LightningMods** [comment]
@chris2677 If that happens just stop the NFS server and restart it, the install will fail but you will be unfrozen, then you can reconnect and try again, WinNFSd is likely the one causing those installations to freeze after X time, I recommend force restarting fuse whenever it freezes with that being said I can't see any possible way the PS4 itself would be the issue rather that again would be the NFS server, if you want the most stability I'd use Windows server 2019's built-in NFS features or any other OS with native NFS features

**Comment by chris2677** [comment]
Thanks I'll give that a try, do you know of any other app I can use on
windows 11 instead?

**Comment by LightningMods** [comment]
@chris2677 there are no better free and open source alternatives I know of,  just paid NFS solutions, Unix systems like MacOS and Linux usually have built-in NFS features you have to enable as NFS is mainly used by Unix systems which Windows is not

**Comment by rootk1t** [comment]
I recently has the following issue with Itemzflow.

I moved (Move to USB0 ->Move App/Game) installed game to external hdd, it booted ok and was playable.
After some time i decided to restore (Restore from USB -> Restore App/Game) game back to internal HDD.
Itemzflow warned me, that "This App is already installed do you want to reinstall it? (All Game data will be deleted!)".
I confirmed. After that indicator of status appeared and stayed on 0%.
I waited for several hours, to be on the safe side, but nothing happened.
I unplugged external HDD and rebooted PS4.
After reboot, i saw ALL my game SAVES disappeared from ps4 storage settings.
I rebooted once more, thinking of it as a temporary issue.
But unfortunately it persisted. Via ftp a checked user\home\XXXXXXXX\savedata\ folder and it was completely empty!
I checked external HDD for save files, but they were not there too(
Also after reboot itemzflow showed like some of my other games, which i never moved, are located on external HDD, though it was not true. To be on the safe side, i had to delete those games.
I don't know exactly what happened, but seems that itemzflow corrupted some .db or profile files which are related to saves.

**Comment by LightningMods** [comment]
@rootk1t when Itemzflow said "All Game data will be deleted!" It meant the game saves too because when itemzflow moves games back to the HDD it has to uninstall the current game and install the PKG again and why it shows that warning message

**Comment by rootk1t** [comment]
> @rootk1t when Itemzflow said "All Game data will be deleted!" It meant the game saves too because when itemzflow moves games back to the HDD it has to uninstall the current game and install the PKG again and why it shows that warning message

I understand it. It's ok for me to loose saves for the game i moved (back and forth).
But why did i loose all game&app saves as a result? Is it normal behaviour?

**Comment by LightningMods** [comment]
@rootk1t I have no idea why you would loose all your saves, Itemzflow itself doesn't even delete the game save when moving back to the HDD it's just what the ps4 does when the app is uninstalled using the internal system function, it's all handled by the ps4

**Comment by rootk1t** [comment]
> @rootk1t I have no idea why you would loose all your saves, Itemzflow itself doesn't even delete the game save when moving back to the HDD it's just what the ps4 does when the app is uninstalled using the internal system function, it's all handled by the ps4
> 
> It could be possible that your hdd or ssd could have gotten corrupted, that would explain it getting stuck at 0% (which is also controlled by the ps4)

Maybe you're right about hdd issue.

**Comment by Antibodyarmy** [comment]
I don't know if this is the issue, but I ran into this with a tplink router with ps4 pkg sender and remote package sender and large size FTP transfers with tons of small files . with my particular tplink router in addition to allowing firewall rules in windows, I needed to disable the LAN ping firewall setting, it would kill a transfer occasionally break the initiating for a little bit, allow it again, then break it again. So now as a precaution i just turn that off before initiating any big FTP transfers and have not had issues since. I even had this issue when connecting my ps4 through my computers 2nd ethernet port it took my router going out from a power outage to find the source of my kind of weird issue.
